std/haxe/async/Callback.hx

```diff
@@ -0,0 +1,65 @@
+package haxe.async;
+
+import haxe.Error;
+import haxe.NoData;
+
+typedef CallbackData<T> = (?error:Error, ?result:T) -> Void;
+
+/**
+	A callback. All callbacks in the standard library are functions which accept
+	two arguments: an error (`haxe.Error`) and a result (`T`). If error is 
+	non-`null`, result must be `null`. The callback type is declared in 	`CallbackData`.
+
+	This abstract defines multiple `@:from` conversions to improve readability of
+	callback code.
+**/
+@:callable
+abstract Callback<T>(CallbackData<T>) from CallbackData<T> {
+	/**
+		Returns a callback of the same type as `cb` which is guaranteed to be
+		non-`null`. If `cb` is given and is not `null` it is returned directly.
+		If `cb` is `null` a dummy callback which does nothing is returned instead.
+	**/
+	public static function nonNull<T>(?cb:Callback<T>):Callback<T> {
+		if (cb == null)
+			return (_, _) -> {};
+		return cb;
+	}
+
+	/**
+		Wraps a function which takes a single optional `haxe.Error` argument into
+		a callback of type `Callback<NoData>`. Allows:
+
+		```haxe
+		var cb:Callback<NoData> = (?err) -> trace("error!", err);
+		```
+	**/
+	@:from public static inline function fromOptionalErrorOnly(f:(?error:Error) -> Void):Callback<NoData> {
+		return (?err:Error, ?result:NoData) -> f(err);
+	}
+
+	/**
+		Wraps a function which takes a single `haxe.Error` argument into a callback
+		of type `Callback<NoData>`. Allows:
+
+		```haxe
+		var cb:Callback<NoData> = (err) -> trace("error!", err);
+		```
+	**/
+	@:from public static inline function fromErrorOnly(f:(error:Error) -> Void):Callback<NoData> {
```

Comment:
I think this and `fromErrorResult` should not exist. It's a source for NPEs.

Reply:
Could you give an example?

Reply:
Sure
```haxe
var cb:Callback<String> = function(error:Error) { // It's clear `null` is not expected here
  switch(error.type) {...} // null pointer error
}
successOp(cb);
```
```haxe
var cb:Callback<String> = function(_, result:String) { // It's clear `null` is not expected here
  trace(result.length); // null pointer error
}
failOp(cb);
```
And it defeats null safety feature.

Reply:
I don't think the first would compile, since `fromErrorOnly` only produces a `Callback<NoData>`.

I would argue the second is a user error. Subverting null safety sounds more serious though…

Reply:
> I don't think the first would compile, since fromErrorOnly only produces a Callback<NoData>.

Data type does not matter in that example. It's still NPE even with `Callback<NoData>`. Maybe implementation should be changed to `if(err != null) f(err)`.

> I would argue the second is a user error.

That's a user error which could never happen if `Null<>` is respected by the API ;)